src/allmydata/mutable/publish.py

```diff
@@ -945,6 +945,17 @@ def update_goal(self):
             serverid = server.get_serverid()
             if server in self.bad_servers:
                 continue
+            # if we have >= 1 grid-managers, this checks that we have
+            # a valid certificate for this server
+            if not server.upload_permitted():
+                self.log(
+                    "No valid grid-manager certificates for '{}' while choosing slots for mutable".format(
+                        server.get_serverid(),
+                    ),
+                    level=log.UNUSUAL,
```

Comment:
Sure ... I think the diff is telling me this is the only logging that got added (well, no other `self.log` statements anyway). Since eliot works with context-managers only, is it "better" to have this function return True or throw an exception when upload is not permitted? (It's probably better to have the above error-message originate inside the `upload_permitted` method anyway -- as you point out elsewhere, we might have other reasons besides grid-manager to not upload to a server).

Reply:
Okay, I tried a thing based on eliot.